File: api-node/src/security/utils/getJWTToken.ts
import bcrypt from 'bcrypt';
import jsonwebtoken from 'jsonwebtoken';
import config from '../../config';
import findUserByEmail from '../../dataSources/postgres/utils/findUserByEmail';

const getJWTToken = async (email: string, password: string): Promise<string> => {
  const user = await findUserByEmail(email);
  if (user == null) {
    throw new Error('No user with this email.');
  }

  const isPasswordValid = await bcrypt.compare(password, user.password);
  if (!isPasswordValid) {
    throw new Error('Incorrect password.');
  }

  return jsonwebtoken.sign({ id: user.id, email: user.email }, config.jwtSecret, { expiresIn: '1d' });
};

export default getJWTToken;

File: api-go/internal/utils/get_jwt_token.go
package utils

import (
	"github.com/golang-jwt/jwt"
	"github.com/rs/zerolog/log"
	"time"
)

type JWT struct {
	JWTToken string `json:"jwtToken"`
}

func GetJWTToken(email string, password string) (JWT, error) {
	var config = GetConfig()

	uid := GetUIDByEmail(email)
	if uid == "" {
		return JWT{}, nil
	}
	isPasswordValid := VerifyPassword(uid, password)
	if !isPasswordValid {
		return JWT{}, nil
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"id":  uid,
		"exp": time.Now().Add(time.Hour * 24).Unix(),
	})

	tokenString, err := token.SignedString([]byte(config.JWTSecret))
	if err != nil {
		log.Error().Err(err).Msg("token.SignedString")
	}

	return JWT{
		JWTToken: tokenString,
	}, nil
}
